src/Streamly/Internal/Data/Array/Foreign.hs

```diff
@@ -457,10 +462,34 @@ getIndex arr i =
 --       in Unfold.lmap f (getIndicesFromThenTo i (i - 1) 0)
 -- @
 --
--- /Unimplemented/
+-- /Pre-release/
+
+{-# INLINE getIndices' #-}
+getIndices' :: (MonadIO m, Storable a) =>
+    D.Stream m Int -> Unfold m (Array a) a
+getIndices' (D.Stream stepi sti) = Unfold step inject
+
+    where
+
+    inject arr = return $ IndexState arr sti
+
+    {-# INLINE_LATE step #-}
+    step (IndexState arr st) = do
+        r <- stepi defState st
+        case r of
+            D.Yield i s -> do
+                case getIndex arr i of
+                    Just x -> return $ D.Yield x (IndexState arr s)
+                    Nothing -> return D.Stop
+            D.Skip s -> return $ D.Skip (IndexState arr s)
+            D.Stop -> do
+                liftIO $ touch (arrContents arr)
+                return D.Stop
+
 {-# INLINE getIndices #-}
-getIndices :: Unfold m (Array a) Int -> Unfold m (Array a) a
-getIndices = undefined
+getIndices :: (MonadIO m, Storable a)
+    => SerialT m Int -> Unfold m (Array a) a
+getIndices (SerialT stream) = getIndices' $ D.fromStreamK stream
```

Comment:
Just use unsafeFreeze/unsafeThaw to convert mutable/immutable arrays, do not duplicate the code.